ociclient/client.go

```diff
@@ -139,11 +141,169 @@ func (c *client) GetManifest(ctx context.Context, ref string) (*ocispecv1.Manife
 		return nil, err
 	}
 
-	var manifest ocispecv1.Manifest
-	if err := json.Unmarshal(data.Bytes(), &manifest); err != nil {
+	var ociArtifact *oci.Artifact
+
+	if desc.MediaType == ocispecv1.MediaTypeImageIndex || desc.MediaType == images.MediaTypeDockerSchema2ManifestList {
+		var index ocispecv1.Index
+		if err := json.Unmarshal(data.Bytes(), &index); err != nil {
+			return nil, err
+		}
+
+		i := oci.Index{
+			Manifests:   []*oci.Manifest{},
+			Annotations: index.Annotations,
+		}
+
+		ociArtifact = oci.NewIndexArtifact(&i)
+
+		for _, mdesc := range index.Manifests {
+			data := bytes.NewBuffer([]byte{})
+			if err := c.Fetch(ctx, ref, mdesc, data); err != nil {
+				return nil, err
+			}
+
+			var manifest ocispecv1.Manifest
+			if err := json.Unmarshal(data.Bytes(), &manifest); err != nil {
+				return nil, err
+			}
+
+			m := oci.Manifest{
+				Desc: mdesc,
+				Data: &manifest,
+			}
+
+			ociArtifact.GetIndex().Manifests = append(ociArtifact.GetIndex().Manifests, &m)
+		}
+	} else if desc.MediaType == ocispecv1.MediaTypeImageManifest || desc.MediaType == images.MediaTypeDockerSchema2Manifest {
+		var manifest ocispecv1.Manifest
+		if err := json.Unmarshal(data.Bytes(), &manifest); err != nil {
+			return nil, err
+		}
+
+		m := oci.Manifest{
+			Desc: desc,
+			Data: &manifest,
+		}
+
+		ociArtifact = oci.NewManifestArtifact(&m)
+	} else {
+		return nil, fmt.Errorf("unable to handle mediatype: %s", desc.MediaType)
+	}
+
+	return ociArtifact, nil
+}
+
+func (c *client) PushOCIArtifact(ctx context.Context, ref string, artifact *oci.Artifact, options ...PushOption) error {
+	opts := &PushOptions{}
+	opts.Store = c.cache
+	opts.ApplyOptions(options)
+
+	tempCache := c.cache
+	if tempCache == nil {
+		tempCache = cache.NewInMemoryCache()
+	}
+
+	resolver, err := c.getResolverForRef(ctx, ref, transport.PushScope)
+	if err != nil {
+		return err
+	}
+	pusher, err := resolver.Pusher(ctx, ref)
+	if err != nil {
+		return err
+	}
+
+	if artifact.IsManifest() {
+		_, err := c.pushManifest(ctx, artifact.GetManifest().Data, pusher, tempCache, opts)
+		return err
+	} else if artifact.IsIndex() {
+		return c.pushImageIndex(ctx, artifact, pusher, tempCache, opts)
+	} else {
+		return errors.New("artifact doesn't contain any manifests")
+	}
```

Comment:
it shouldn't happen, nevertheless: I would keep the `else` branch here, otherwise it would be a bad smell in my opinion because we're not handling all theoretically possible condition branches. I've updated the error handling to print out the complete oci artifact in JSON if we run into this branch.

Reply:
yes thats fine. But we could maybe panic here as we are in an inconsistent state. WDYT?

Reply:
yes, a panic might be even more appropriate. I'll include it